### PR TITLE
Neoways features framework blog post

### LINKED ISSUE #8: Write a project post
Our currently open source softwares projects are (to name a few of them):

+ nash
+ neosearch
+ klb
+ waveparser
+ signal
+ drm

Pick and write a project post in `_projects`. Use something like this as header:

```
---
title: "Oatmeal Cookies"
image: 
  path: /images/oatmeal-cookies-lg.jpg
  thumbnail: /images/oatmeal-cookies-400x200.jpg
  caption: "Photo from [Pexels](https://www.pexels.com)"
layout: page
---
```

### PR DESCRIPTION
**Pull request by mfrata**
Blog post about features framework!

@ryukinix 

Closes #8

### CODE REVIEW

File: _posts/2019-08-14-neoways-feature-framework.md [hunk 1]
```diff
@@ -0,0 +1,203 @@
...
+</figure> 
+
+
+  The framework is cloud-agnostic but it currently requires provisioning of cluster management and other supporting tools for scheduling and serving. These tools, generally, are offered by a cloud provider. They can simplify the task of infrastructure handling.
+
+# The main engine
+
+The main component of the framework is the **Features Builder**. It is an extensible framework for features development, where they are logically organized by domains with a specific primary key.
+
+<figure class="align-center">
```
Comment (ryukinix) [inline _posts/2019-08-14-neoways-feature-framework.md]: Cool.

File: _posts/2019-08-14-neoways-feature-framework.md [hunk 2]
```diff
@@ -0,0 +1,203 @@
...
+  thumbnail: /images/posts/nw-feats-framework/blog_poster.png
+categories:
+  - data-science
+tags:
+  - feature-engineering
+  - pyspark
+excerpt: "Scalling data science development"
+---
+
+As a company grows it becomes more difficult to maintain all of its data science projects. The way the features are generated across these projects can introduce some problems such as the following: duplicates, lack of features consistency, absence of documentation, or even scalability issues. Have you ever heard of two teams using the same features on different problems while being calculated differently? Or a data scientist having to discard a whole Feature Engineering procedure due to the cumbersomeness do adapt to other projects?
```
Comment (fcpenha) [inline _posts/2019-08-14-neoways-feature-framework.md]: to adapt

**Comment by mfrata** [comment]
> There is this path: `/images/posts/nw-feats-framework` where you put the images.
> The last dirname is the post name, however, you create the post with other name: `_posts/2019-08-14-neoways-feature-framework.md`
> 
> So there is two names for the same post:
> 
>     * neoways-feature-framework (which will generates the url after jekyll build)
> 
>     * nw-feats-framework
> 
> 
> I suggest you to choose only one name for both cases. To make this change simpler, I think renaming the markdown file to `_posts/2019-08-14-nw-feats-framework.md` it's better, that way you will avoid replacing the paths of images in the post.

Done!

File: _posts/2019-08-14-neoways-feature-framework.md [hunk 3]
```diff
@@ -0,0 +1,203 @@
+---
+author: mfrata 
+title: "Neoway's Feature Framework"
+image:
+  path: /images/posts/nw-feats-framework/blog_poster.png
```
Comment (ryukinix) [inline _posts/2019-08-14-neoways-feature-framework.md]: I think you should update this path to `/images/posts/neoways-feature-framework`

File: _posts/2019-08-14-neoways-feature-framework.md [hunk 4]
```diff
@@ -0,0 +1,203 @@
...
+  thumbnail: /images/posts/nw-feats-framework/blog_poster.png
```
Comment (ryukinix) [inline _posts/2019-08-14-neoways-feature-framework.md]: I think you should update this path to `/images/posts/neoways-feature-framework`

File: _posts/2019-08-14-neoways-feature-framework.md [hunk 5]
```diff
@@ -0,0 +1,203 @@
...
+
+In this blog post, we will dive deep into the internals of the **Neoway’s Features Framework**, explaining all of its components and discussing the outcomes for our company.
+
+# Where the features live
+
+There are more than 700 entities that are processed periodically and are stored on Neoway’s data lake. The entities that are going to be used by the framework’s calculation are converted to the parquet file format for better operability, space usage, and querying efficiency. The generated features go back to the data lake in a parquet file format.
+
+
+<figure class="align-center">
+  <img src="{{ '/images/posts/nw-feats-framework/features_store.png' | absolute_url }}" alt="">
```
Comment (ryukinix) [inline _posts/2019-08-14-neoways-feature-framework.md]: I think you should update this path to `/images/posts/neoways-feature-framework`

File: _posts/2019-08-14-neoways-feature-framework.md [hunk 6]
```diff
@@ -0,0 +1,203 @@
...
+
+
+  The framework is cloud-agnostic but it currently requires provisioning of cluster management and other supporting tools for scheduling and serving. These tools, generally, are offered by a cloud provider. They can simplify the task of infrastructure handling.
+
+# The main engine
+
+The main component of the framework is the **Features Builder**. It is an extensible framework for features development, where they are logically organized by domains with a specific primary key.
+
+<figure class="align-center">
+  <img src="{{ '/images/posts/nw-feats-framework/features_builder.png' | absolute_url }}" alt="">
```
Comment (ryukinix) [inline _posts/2019-08-14-neoways-feature-framework.md]: I think you should update this path to `/images/posts/neoways-feature-framework`

File: _posts/2019-08-14-neoways-feature-framework.md [hunk 7]
```diff
@@ -0,0 +1,203 @@
...
+
+  Another advantage of this way of feature development is something that we call: **Backward Feature Generation**. It is a primordial "feature" of the Features Framework (pun intended). Whenever a new feature comes, its calculation not only occurs on the current timestamp but also for past dates whenever historical data are available. Using the same example described above, if our `feature_one` is created in October but the data source exists since January, the feature will be extended to those dates. In essence, this means that `feature_one` is available for multiple dates starting from January, even if its conception only happens in October. This capability is great as it allows us to backtest our models.
+
+
+# Can we make serving easier?
+
+The real value of the Builder and the Store only comes when our stakeholders are making use of the features. However, accessing a storage bucket on the cloud is not the best way to achieve that. To solve this, we developed a high-level layer to access those: the **Features Selector**.
+
+<figure class="align-center">
+  <img src="{{ '/images/posts/nw-feats-framework/features_selector.png' | absolute_url }}" alt="">
```
Comment (ryukinix) [inline _posts/2019-08-14-neoways-feature-framework.md]: I think you should update this path to `/images/posts/neoways-feature-framework`

File: _posts/2019-08-14-neoways-feature-framework.md [hunk 8]
```diff
@@ -0,0 +1,203 @@
...
+...      .select('pkey', 'feature_one', ..., 'feature_N'))
+{% endhighlight %}
+
+
+# How about turning it into a product?
+
+At first, the Features Framework was aimed to solve our internal scalability challenges. As it grew we saw the potential that it could bring also to our clients. We proved our hypothesis using the Selector to deliver the features to our customers. Nonetheless, someone still had to stop and do the job of generating them. Meaning that it would not scale well. That led to our external way of serving it: the **Features API**.
+
+<figure class="align-center">
+  <img src="{{ '/images/posts/nw-feats-framework/features_api.png' | absolute_url }}" alt="">
```
Comment (ryukinix) [inline _posts/2019-08-14-neoways-feature-framework.md]: I think you should update this path to `/images/posts/neoways-feature-framework`

File: _posts/2019-08-14-neoways-feature-framework.md [hunk 9]
```diff
@@ -0,0 +1,203 @@
...
+
+The API is a service written in Golang that provides a low latency interface for features consumption and also has built-in security and billing capabilities provided by our Platform team. A key aspect is that we only include features that we use and have a proven value in real world applications. As it turns out, those features are also bringing results to our clients in many segments such as finance, insurance, recovery, retail, consumer goods and many others. Before requesting the features, the client can access its metadata as well as the timestamps available. 
+
+
+# Let’s talk about data quality
+
+Until now we saw all the value of the Features Framework can create. But how do we ensure quality? You’ve guessed it: **Features Monitor**.
+
+<figure class="align-center">
+  <img src="{{ '/images/posts/nw-feats-framework/features_monitor.png' | absolute_url }}" alt="">
```
Comment (ryukinix) [inline _posts/2019-08-14-neoways-feature-framework.md]: I think you should update this path to `/images/posts/neoways-feature-framework`

File: _posts/2019-08-14-neoways-feature-framework.md [hunk 10]
```diff
@@ -0,0 +1,203 @@
...
+      ...
+    },
+    ...
+    ]
+  }
+}
+{% endhighlight %}
+
+<figure class="align-center">
+  <figcaption>Example of Features Monitor report.</figcaption>
```
Comment (ryukinix) [inline _posts/2019-08-14-neoways-feature-framework.md]: What figure is that? I didn't saw the image source.

File: _posts/2019-08-14-neoways-feature-framework.md [hunk 11]
```diff
@@ -0,0 +1,203 @@
...
+</figure> 
+
+
+# Le grand orchestrateur
+
+All of the components above are crucial parts of the framework. Yet there is another hidden component that plays a big role in the features’ calculations. We’re talking about the orchestrator, the scheduler. We’re talking about [Apache Airflow](https://airflow.apache.org/).
+
+<figure class="align-center">
+  <img src="{{ '/images/posts/nw-feats-framework/features_framework.png' | absolute_url }}" alt="">
```
Comment (ryukinix) [inline _posts/2019-08-14-neoways-feature-framework.md]: I think you should update this path to `/images/posts/neoways-feature-framework`

File: _posts/2019-08-14-neoways-feature-framework.md [hunk 12]
```diff
@@ -0,0 +1,203 @@
...
+
+So should you invest in your Features Framework at your company? It depends on the stage you are at. But if you wish to scale up it is a good idea to invest in a solution to improve the feature development. And the Features Framework has a great return on investment. What do you think about our Features Framework? Leave us a message below or drop us a line.
+
+
+*Special thanks to all people involved on the development of this framework: [Andre Boechat](https://www.linkedin.com/in/boechat107/), [Breno Costa](https://www.linkedin.com/in/breno-c-costa/), [Gabriel Alvim](https://www.linkedin.com/in/gabriel-benatti-alvim-520b69a2/), [Igor Cortez](https://www.linkedin.com/in/igor-cortez-56793825/), [Luis Pelison](https://www.linkedin.com/in/luis-felipe-pelison-243791107/), and [Manoel Vilela](https://www.linkedin.com/in/lerax/).*
```
Comment (ryukinix) [inline _posts/2019-08-14-neoways-feature-framework.md]: Nice to see that references here. Thanks.

File: _posts/2019-08-14-neoways-feature-framework.md (hunk 10 above)
Comment (mfrata) [reply _posts/2019-08-14-neoways-feature-framework.md]: Just the caption for the code example

**Comment by mfrata** [comment]
@ryukinix I updated the images paths. However I can't see the images, they're not rendering.

image

Can you fetch the changes and check it out?

**Comment by ryukinix** [comment]
In my machine the result is worse: 
Screenshot from 2019-08-16 13-36-49


Seems something related to the new change on `docker run` and volumes.

**Comment by ryukinix** [comment]
Well, ignoring docker volume stuff by calling: `make image; make blog DOCKER_VOLUME_ARGS=''`

I got the same result you have here. One thing that I note trying to debug was: local image paths are been translated to public urls with neowaylabs.github.io as prefix. This doesn't works obviously because we are running local.  See by yourself:
Screenshot from 2019-08-16 14-02-01


But changing the `src` of image from `https://neowaylabs.github.io/images/posts/neoways-feature-framework/features_store.png` to `/images/posts/neoways-feature-framework/features_store.png` the picture is displayed. 

Screenshot from 2019-08-16 14-03-34

**Comment by ryukinix** [comment]
I think if we merge this as it is will work, but would be good understanding how make this works correctly in a local server too.

**Comment by ryukinix** [comment]
Okay, I found a potential solution.

Instead of: 
```html
<img src="{{ '/images/posts/neoways-feature-framework/features_store.png' | absolute_url }}" alt="">
```

Use this
```html
<img src="/images/posts/neoways-feature-framework/features_store.png" alt="">
```

At least locally this works ok. All the images is rendered without problems.

**Comment by ryukinix** [comment]
I add a commit fixing the paths @mfrata. Try to run now in your machine.

**Comment by mfrata** [comment]
@ryukinix It worked. You can Merge It!!!

**Pull request merged**